### PR TITLE
Seting a TTL for storing terminated pods in etcd

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/GoogleCloudPlatform/kubernetes/issues/7660

### CODE REVIEW

**Comment by nikhiljindal** [comment]
cc @lavalamp

**Comment by thockin** [comment]
Does this mean that a node dying does not need to delete pods?  YAY!

File: pkg/registry/pod/etcd/etcd.go [hunk 1]
```diff
@@ -39,6 +39,9 @@ import (
 	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/fielderrors"
 )
 
+// Clear terminated pods from etcd after 4 days.
+var ttlForTerminatedPods uint64 = 4 * 24 * 3600
```
Comment (derekwaynecarr) [inline pkg/registry/pod/etcd/etcd.go]: Can we make this configurable?  

There are some pods that are succeed/failed that we may want to keep longer than others.

@smarterclayton - thoughts?

**Comment by derekwaynecarr** [comment]
So I thought we looked at TTL for and found issues with etcd having a large number of nodes with TTL as part of this work, and ultimately held off using it because of stress test concerns:

https://github.com/GoogleCloudPlatform/kubernetes/pull/9165

My concern is that we will have over 4 days the potential for a very large number of pods in success/failure state each with a TTL applied especially when Jobs are added to the server.

So over 4 day period, how many TTLs do you think we could have in etcd and what kind of impact does that have on the system?

Is there an advantage to having a timestamp when a pod succeed/fails and writing a more typical controller pattern to find pods that completed N days ago (where N is configurable)?

**Comment by smarterclayton** [comment]
We did - that's why graceful deletion no longer uses them and instead uses
rectifier loops.

We should be using rectifier loops for this as well.  We need to have a
rectifier for pods on deleted nodes - handling this as part of that is a
better direction.

On Jul 30, 2015, at 8:44 PM, Derek Carr <notifications@github.com> wrote:

So I thought we looked at TTL for and found issues with etcd having a large
number of nodes with TTL as part of this work, and ultimately held off
using it because of stress test concerns:

#9165 <https://github.com/GoogleCloudPlatform/kubernetes/pull/9165>

My concern is that we will have over 4 days the potential for a very large
number of pods in success/failure state each with a TTL applied especially
when Jobs are added to the server.

So over 4 day period, how many TTLs do you think we could have in etcd and
what kind of impact does that have on the system?

Is there an advantage to having a timestamp when a pod succeed/fails and
writing a more typical controller pattern to find pods that completed N
days ago (where N is configurable)?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12055#issuecomment-126531657>
.

**Comment by derekwaynecarr** [comment]
We are already running a number of short lived pods to help drive builds
and deployment hooks in OpenShift, need to make a note to see impact there.

@smarterclayton - do u recall if we reap any of those yet?

**Comment by nikhiljindal** [comment]
oh. Didnt know about that PR.
Thanks @derekwaynecarr for pointing me to it.
Will take a look.

**Comment by smarterclayton** [comment]
We don't - nikhil, let's catch up tomorrow - we probably need to spec a pod cleaner anyway and if it's not in graceful delete it's probably good to do afterwards.

**Comment by lavalamp** [comment]
This TTL mechanism is not appropriate for cleaning up after a job
controller, I agree. We don't have a job controller yet, though...

**Comment by smarterclayton** [comment]
OpenShift is using regular pods for jobs already (as a workload on Kube)
and I think they would have a high pod fraction in some of our deployments.

TTL is not terribly efficient in etcd - it causes latency spikes under
write load and also ties us to etcd more strongly.  The one advantage of
etcd is that it relies on a central timekeeper (ish) and so is not subject
to clock issues.

I propose (for several reasons) that we need a Pod Termination Controller
for the purpose of ensuring that terminated (graceful or otherwise) pods
are cleaned up from the system.

The invariants this controller would maintain:

* pods that scheduled to nodes that do not exist are deleted immediately.
* pods that were gracefully deleted on a node that does not exist are fully
deleted in some period of time TerminationGracePeriodSeconds +
MinimumSafeClockSkewEstimation
* pods that have an active termination deadline are eventually gracefully
deleted in some period of time ActiveDeadlineSeconds +
MinimumSafeClockSkewEstimation
* pods that are completed are eventually cleaned up

The node controller would attempt to maintain the invariants:

* pods on nodes that are not responding are gracefully deleted
* nodes that are not responding are eventually deleted

There's probably other invariants, but these to me seem to overlap, not be
disjoint.

Alternatively, the Apiserver could do read repair on pods that appear to
violate some of the above invariants by completing their deletion.



On Jul 31, 2015, at 1:23 AM, Daniel Smith <notifications@github.com> wrote:

This TTL mechanism is not appropriate for cleaning up after a job
controller, I agree. We don't have a job controller yet, though...
On Jul 30, 2015 5:44 PM, "Derek Carr" <notifications@github.com> wrote:

> So I thought we looked at TTL for and found issues with etcd having a
> large number of nodes with TTL as part of this work, and ultimately held
> off using it because of stress test concerns:
>
> #9165 <https://github.com/GoogleCloudPlatform/kubernetes/pull/9165>
>
> My concern is that we will have over 4 days the potential for a very large
> number of pods in success/failure state each with a TTL applied especially
> when Jobs are added to the server.
>
> So over 4 day period, how many TTLs do you think we could have in etcd and
> what kind of impact does that have on the system?
>
> Is there an advantage to having a timestamp when a pod succeed/fails and
> writing a more typical controller pattern to find pods that completed N
> days ago (where N is configurable)?
>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/GoogleCloudPlatform/kubernetes/pull/12055#issuecomment-126531657
>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12055#issuecomment-126569894>
.

**Comment by nikhiljindal** [comment]
Will a generic controller to delete all "marked-for-deletion" resources be better than a specific "Pod Termination Controller"?

I am trying to come up with scenarios in which we will need this GC controller. I can see the following cases in which we want to delete pods:
* Pod was running on a node, which was stopped by the user: In this case, nodeController will call apiserver to delete all pods running on that node. pod will be cleared from etcd as a result. In its sync loop, kubelet will detect that it is running a pod that is not in etcd and will then gracefully kill the running pod and its containers.
Hence, we do not need a GC controller in this case.
* Pod terminated (failed/succeeded): Do we always want to delete the pod from etcd whenever it reaches terminated state? If yes, then can the apiserver just delete the pod, whenever it gets a status update request, to update the status of the pod to any of the terminated state?
Then we wont need a GC controller for pods.

What am I missing?

**Comment by davidopp** [comment]
> The invariants this controller would maintain:
>
> * pods that scheduled to nodes that do not exist are deleted immediately.
> * pods that were gracefully deleted on a node that does not exist are fully
> deleted in some period of time TerminationGracePeriodSeconds +
> MinimumSafeClockSkewEstimation
> * pods that have an active termination deadline are eventually gracefully
> deleted in some period of time ActiveDeadlineSeconds +
> MinimumSafeClockSkewEstimation
> * pods that are completed are eventually cleaned up

I wonder if the last one should be handled instead by the controller that scheduled the Pod (at the cost of some possibly duplicated logic across controllers). I suspect we may want to vary this policy across controllers (e.g. batch vs. replication)

>
> The node controller would attempt to maintain the invariants:
> 
> * pods on nodes that are not responding are gracefully deleted
> * nodes that are not responding are eventually deleted

I don't think we want to delete nodes even when they become non-responsive (not ready), as they might come back. We should only delete them when they are removed from the cloud provider or (for bare metal) by an admin deleting the node object via the API.

**Comment by smarterclayton** [comment]
On Tue, Aug 4, 2015 at 1:33 AM, David Oppenheimer <notifications@github.com>
wrote:

> The invariants this controller would maintain:
>
>    - pods that scheduled to nodes that do not exist are deleted
>    immediately.
>    - pods that were gracefully deleted on a node that does not exist are
>    fully deleted in some period of time TerminationGracePeriodSeconds +
>    MinimumSafeClockSkewEstimation
>    - pods that have an active termination deadline are eventually
>    gracefully deleted in some period of time ActiveDeadlineSeconds +
>    MinimumSafeClockSkewEstimation
>    - pods that are completed are eventually cleaned up
>
> I wonder if the last one should be handled instead by the controller that
> scheduled the Pod (at the cost of some possibly duplicated logic across
> controllers). I suspect we may want to vary this policy across controllers
> (e.g. batch vs. replication)
>

Probably.  Just included it here for completeness.  It depends on whether
those controllers are expected to handle the other invariants on their own.


> The node controller would attempt to maintain the invariants:
>
>    - pods on nodes that are not responding are gracefully deleted
>    - nodes that are not responding are eventually deleted
>
> I don't think we want to delete nodes even when they become non-responsive
> (not ready), as they might come back. We should only delete them when they
> are removed from the cloud provider or (for bare metal) by an admin
> deleting the node object via the API.
>

I agree - substitute delete for "marked as not available".  I don't know
the context behind the current node controller design, but I would assume
we should *not* be deleting the node, just marking it unreachable (in which
case the invariant is unreachable nodes have their pods evacuated)

**Comment by smarterclayton** [comment]
On Tue, Aug 4, 2015 at 1:30 AM, Nikhil Jindal <notifications@github.com>
wrote:

> Will a generic controller to delete all "marked-for-deletion" resources be
> better than a specific "Pod Termination Controller"?
>
> I am trying to come up with scenarios in which we will need this GC
> controller. I can see the following cases in which we want to delete pods:
>
>    - Pod was running on a node, which was stopped by the user: In this
>    case, nodeController will call apiserver to delete all pods running on that
>    node. pod will be cleared from etcd as a result. In its sync loop, kubelet
>    will detect that it is running a pod that is not in etcd and will then
>    gracefully kill the running pod and its containers. Hence, we do not need a
>    GC controller in this case.
>
>
Kubelet dies and never comes back.  Kubelet can't be the only thing that
ensures gracefully stopped pods get deleted from etcd.



>
>    - Pod terminated (failed/succeeded): Do we always want to delete the
>    pod from etcd whenever it reaches terminated state? If yes, then can the
>    apiserver just delete the pod, whenever it gets a status update request, to
>    update the status of the pod to any of the terminated state? Then we wont
>    need a GC controller for pods.
>
>
There is no guarantee the kubelet would ever send the status update that
says it observed the deletion.  If the node is ballistically decomissioned
there will never be another status update.

**Comment by nikhiljindal** [comment]
On Tue, Aug 4, 2015 at 9:13 AM, Clayton Coleman <notifications@github.com>
wrote:

> On Tue, Aug 4, 2015 at 1:30 AM, Nikhil Jindal <notifications@github.com>
> wrote:
>
> > Will a generic controller to delete all "marked-for-deletion" resources
> be
> > better than a specific "Pod Termination Controller"?
> >
> > I am trying to come up with scenarios in which we will need this GC
> > controller. I can see the following cases in which we want to delete
> pods:
> >
> > - Pod was running on a node, which was stopped by the user: In this
> > case, nodeController will call apiserver to delete all pods running on
> that
> > node. pod will be cleared from etcd as a result. In its sync loop,
> kubelet
> > will detect that it is running a pod that is not in etcd and will then
> > gracefully kill the running pod and its containers. Hence, we do not
> need a
> > GC controller in this case.
> >
> >
> Kubelet dies and never comes back. Kubelet can't be the only thing that
> ensures gracefully stopped pods get deleted from etcd.
>

There are 2 ways this can play out:
1. Node controller calls delete on apiserver, which clears the pod from
etcd. Later, kubelet realizes that it is running a pod that is not in etcd
and then deletes it (gracefully or not).
Questions: Is it fine to delete the pod from etcd before it is deleted by
kubelet?
Can kubelet gracefully delete the pod without having to read it from etcd?
@yujuhong ?
2. Node controller just marks the pod to be deleted. kubelet realizes that
it is running a pod that is marked for deletion and hence deletes it. When
its done, it calls delete pod on apiserver which clears the pod from etcd.

We wont need a GC if we support 1.


>
>
> >
> > - Pod terminated (failed/succeeded): Do we always want to delete the
> > pod from etcd whenever it reaches terminated state? If yes, then can the
> > apiserver just delete the pod, whenever it gets a status update request,
> to
> > update the status of the pod to any of the terminated state? Then we wont
> > need a GC controller for pods.
> >
> >
> There is no guarantee the kubelet would ever send the status update that
> says it observed the deletion. If the node is ballistically decomissioned
> there will never be another status update.


Wont NodeController handle that case? If the node is up and running then
kubelet will handle deletion. If not, then the node controller will ensure
that it has deleted pods assigned to unhealthy nodes.

**Comment by nikhiljindal** [comment]
cc @bgrant0607

**Comment by smarterclayton** [comment]
On Aug 4, 2015, at 2:41 PM, Nikhil Jindal <notifications@github.com> wrote:

On Tue, Aug 4, 2015 at 9:13 AM, Clayton Coleman <notifications@github.com>
wrote:

> On Tue, Aug 4, 2015 at 1:30 AM, Nikhil Jindal <notifications@github.com>
> wrote:
>
> > Will a generic controller to delete all "marked-for-deletion" resources
> be
> > better than a specific "Pod Termination Controller"?
> >
> > I am trying to come up with scenarios in which we will need this GC
> > controller. I can see the following cases in which we want to delete
> pods:
> >
> > - Pod was running on a node, which was stopped by the user: In this
> > case, nodeController will call apiserver to delete all pods running on
> that
> > node. pod will be cleared from etcd as a result. In its sync loop,
> kubelet
> > will detect that it is running a pod that is not in etcd and will then
> > gracefully kill the running pod and its containers. Hence, we do not
> need a
> > GC controller in this case.
> >
> >
> Kubelet dies and never comes back. Kubelet can't be the only thing that
> ensures gracefully stopped pods get deleted from etcd.
>

There are 2 ways this can play out:
1. Node controller calls delete on apiserver, which clears the pod from
etcd. Later, kubelet realizes that it is running a pod that is not in etcd
and then deletes it (gracefully or not).
Questions: Is it fine to delete the pod from etcd before it is deleted by
kubelet


Once termination period is exceeded, in theory any component can delete it
(by contract).  We let the kubelet be primary so we aren't affected by
clock skew.


Can kubelet gracefully delete the pod without having to read it from etcd?
@yujuhong ?



Define read it?  You mean, can graceful deletion continue once etcd
receives a record of the deletion?  Yes.  However, the users intended
deletion grace period needs to be recorded to etcd at least once.

Part of the reason why we wait to remove the pod is to allow backpressure
on deletion and to provide valuable feedback to the user (if that's what
you were suggesting)


2. Node controller just marks the pod to be deleted. kubelet realizes that
it is running a pod that is marked for deletion and hence deletes it. When
its done, it calls delete pod on apiserver which clears the pod from etcd.

We wont need a GC if we support 1.


>
>
> >
> > - Pod terminated (failed/succeeded): Do we always want to delete the
> > pod from etcd whenever it reaches terminated state? If yes, then can the
> > apiserver just delete the pod, whenever it gets a status update request,
> to
> > update the status of the pod to any of the terminated state? Then we
wont
> > need a GC controller for pods.
> >
> >
> There is no guarantee the kubelet would ever send the status update that
> says it observed the deletion. If the node is ballistically decomissioned
> there will never be another status update.


Wont NodeController handle that case? If the node is up and running then
kubelet will handle deletion. If not, then the node controller will ensure
that it has deleted pods assigned to unhealthy nodes.


>
> —
> Reply to this email directly or view it on GitHub
> <
https://github.com/GoogleCloudPlatform/kubernetes/pull/12055#issuecomment-127662147
>
> .
>

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/12055#issuecomment-127707448>
.

**Comment by bgrant0607** [comment]
I believe we've decided not to use TTL.

**Pull request closed without merging**